X-Withdrawn/RFC0043-Loading-Module-Into-Isolated-AssemblyLoadContext.md

```diff
@@ -0,0 +1,183 @@
+The `ModuleInfo` object will hold a reference to the load context used for this module.
+
+PowerShell assembly loading code and `Add-Type` needs to be changed to not call `Assembly.LoadFrom`.
+Instead, PowerShell needs to call `LoadFromAssemblyPath` on the corresponding load context instance:
+
+- It queries for the current `EngineSessionState` to see if it's from an isolated module;
+- If yes, then the load context of that module is used, otherwise, default load context `AssemblyLoadContext.Default` is used.
+
+_This means we will have type identity issues within a session_ -- type C and C' are the same type in terms of the fully-qualified type name,
+but they are from two different assembly instances that are loaded into two load contexts.
+This would raise multiple folds of challenges that will be discussed below
+
+### Interoperability
+
+Imagine the assemblies A and B are loaded into ALC-M (isolated module M).
+Assembly A exposes a cmdlet `Demo-Cmdlet` that has a parameter accepting an object of type B::C.
+Let's say a different version of B is loaded into the default-ACL, and the result assembly instance is B'.
+So, from the default-ACL (default Runspace), one can create an object of B'::C' and call `Demo-Cmdlet` with it.
+That call will result in a type casting exception with the confusing error like "cannot cast B::C to B::C".
+
+This is because `B'::C'` and `B::C` are two different types,
+even though they have the same fully-qualified names and maybe members.
+This is the interoperability issue that we will face once PowerShell allow modules to be loaded into different load contexts.
+
+This problem can be even worse when it comes to how `Import-Module M -Isolated` deals with M's nested modules and required modules.
+Unless having all nested and required modules loaded into the same custom load context,
+there could be interoperability issues when M calls cmdlets from its nested modules or required modules.
+
+#### Mitigation by updating Type Resolution
+
+We can mitigate the interoperability issue by enhance the `TypeResolver` to make it aware of the current `EngineSessionState` for a type resolution operation.
+If it happens in the scope of a isolated module,
+then `TypeResolver` first searches assemblies in the associated custom-ALC,
+followed by searching in the default-ALC.
+
+We also need to support a new syntax for the type reference in PowerShell:
+
+```powershell
+[ModuleName\TypeName]
+```
+
+When the `ModuleName` part is specified, PowerShell can check if such a module exists and whether it has a custom load context.
+If it's a module with a custom load context, then the type resolution will search assemblies from the custom-ALC and then the default-ALC to find the `TypeName`.
+Otherwise, the type resolution will search the default-ALC only.
+
+So in the above imagined scenario, it's possible to make it work as expected if you construct the object using `[M\B::C]` from the default-ALC.
```

Comment:
We could use `[M::B::C]`. Using `\` is Windows style and we will have to consider `/`. Although with `::` we could get  difficulties in parser with provider form.